txnats/txnats.py

```diff
@@ -79,58 +87,117 @@ def __init__(self, own_reactor=None, verbose=True, pedantic=False,
+                        raise IncompleteCommandError()
+                elif command == "PONG":
+                    self.pout -= 1
+                    val = data_buf.readline()
+                    if val != b'\r\n':
+                        self.remaining_bytes += command + val
+                        raise IncompleteCommandError()
+                elif command == "INFO":
+                    val = data_buf.readline()
+                    if not val.endswith(b'\r\n'):
+                        self.remaining_bytes += command + val
+                        raise IncompleteCommandError()
+                    settings = json.loads(val)
+                    self.server_settings = ServerInfo(**settings)
+                    log.msg(json.dumps(settings), separators=(',', ':'))
+                    self.status = CONNECTED
+                    self.connect()
+                    if self.on_connect_d:
+                        self.on_connect_d.callback(self)
+                        self.on_connect_d = None
                 else:
-                    reply_to = None
-                sid = int(meta_data[1])
-                if sid in self.sids:
-                    on_msg = self.sids[sid]
-                else:
-                    on_msg = None
-
-                if on_msg:
-                    on_msg(self, sid, subject, reply_to,
-                           data_buf.read(n_bytes))
-                    data_buf.readline()
-                elif self.on_msg:
-                    self.on_msg(self, data_buf.read(n_bytes))
-                    data_buf.readline()
-                else:
-                    stdout.write(data_buf.read(n_bytes))
-                    stdout.write(data_buf.readline())
-            elif command == "PING":
-                self.pong()
-                data_buf.readline()
-            elif command == "PONG":
-                self.pout -= 1
-                data_buf.readline()
-            elif command == "INFO":
-                settings = json.loads(data_buf.read())
-                self.server_settings = ServerInfo(**settings)
-                self.status = CONNECTED
-                self.connect()
-                self.on_connect_d.callback(self)
-            else:
-                log.msg("Not handled command is: {!r}".format(command),
-                        logLevel=logging.DEBUG)
-            if not data_buf.peek(1):
-                log.msg("emptied data",
-                        logLevel=logging.DEBUG)
-                break
+                    log.msg("Not handled command is: {!r}".format(command),
+                            logLevel=logging.DEBUG)
+                    val = data_buf.read()
+                    self.remaining_bytes += command + val
+                if not data_buf.peek(1):
+                    log.msg("emptied data",
+                            logLevel=logging.DEBUG)
+                    break
+        except IncompleteCommandError:
+            pass
```

Comment:
On second thought, I don't think handling this with an exception is the right way, since parsing the protocol across dataReceived is normal.